File: Algorithms/0922_Sort_Array_By_Parity_II/Java/SortArrayByParityIISolution2.java

// Space: O(1)
// Time: O(n)

class Solution {
    public int[] sortArrayByParityII(int[] nums) {
        int even = 0, odd = 1, temp = 0;
        while (even < nums.length && odd < nums.length) {
            if (nums[even] % 2 == 0) {
                even += 2;
            } else if (nums[odd] % 2 != 0) {
                odd += 2;
            } else {
                temp = nums[even];
                nums[even] = nums[odd];
                nums[odd] = temp;
                even += 2;
                odd += 2;
            }
        }
        return nums;
    }
}

File: Algorithms/0922_Sort_Array_By_Parity_II/Python/Sort_Array_By_Parity_II_Solution_2.py

# Space: O(1)
# Time: O(n)

class Solution:
    def sortArrayByParityII(self, nums):
        even, odd = 0, 1
        while even < len(nums) and odd < len(nums):
            if nums[even] % 2 == 0:
                even += 2
            elif nums[odd] % 2 != 0:
                odd += 2
            else:
                nums[odd], nums[even] = nums[even], nums[odd]
                even += 2
                odd += 2
        return nums




